README.md
# RestraintLib
Bond length and angle restraints for DNA and RNA oligonucleotides.

# Citing

    M.Kowiel, D.Brzezinski, M. Gilski, M.Jaskolski (2020).
    Conformation-dependent restraints for polynucleotides: The sugar moiety.
    Nucleic Acids Res. 48, 962–973. https://doi.org/10.1093/nar/gkz1122 OpenAccess

    M.Gilski, J.Zhao, M.Kowiel, D.Brzezinski, D.H.Turner, M.Jaskolski (2019).
    Accurate geometrical restraints for Watson–Crick base pairs.
    Acta Cryst. B75, 235-245. https://doi.org/10.1107/S2052520619002002 OpenAccess

    M.Kowiel, D.Brzezinski, M.Jaskolski (2016).
    Conformation-dependent restraints for polynucleotides: I. clustering of the geometry of the phosphodiester group.
    Nucleic Acids Res. 44, 8479–8489. https://doi.org/10.1093/nar/gkw717 OpenAccess

# Dependencies

Scripts were tested with

* Python 3.7
* cctbx==2021.6 [cctbx_project](https://github.com/cctbx/cctbx_project)
* scikit-learn==1.0.2
* numpy==1.20.3

# Installation

Install cctbx with use of miniconda/anacona setup (Option 1) or manually (Option 2).
It seems cctbx installation from pypi does not work at the moment (2020.11.27).

## Option 1: use anaconda setup

1. Clone repository or download sourcecode.

    ```bash
    git clone https://github.com/mkowiel/restraintlib.git
    ```

2. Download Miniconda: [minicnda](https://docs.conda.io/en/latest/miniconda.html#linux-installers).
You can download for example https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh

3. Install Miniconda

    ```bash
    sh .\Miniconda3-latest-Linux-x86_64.sh
    ```

4. Create environment (in the example with `Python3.7` and the `cctbx` env name).

    ```bash
    conda create -n cctbx python=3.7
    conda install --force-reinstall -n cctbx -c conda-forge cctbx-base
    conda install --force-reinstall -n cctbx -c conda-forge pytest
    conda install --force-reinstall -n cctbx -c conda-forge "scikit-learn>1.0,<1.1"
    conda install -c conda-forge conda-build conda-verify
    ```

4. Install library

    ```bash
    # make sure you build on base environment
    conda activate base
    conda-build . --python 3.7 -c conda-forge
    conda install --force-reinstall -n cctbx -c local --offline restraintlib
    ```

## Option 2: manual installation

Follow the installation instruction in the [cctbx_project](https://github.com/cctbx/cctbx_project)

1. Download [https://raw.githubusercontent.com/cctbx/cctbx_project/master/libtbx/auto_build/bootstrap.py](https://raw.githubusercontent.com/cctbx/cctbx_project/master/libtbx/auto_build/bootstrap.py) in the directory where the cctbx and its dependencies shall be installed.
2. On Linux or Mac OS execute it: python bootstrap.py (you may want to run it with the --help option first to discover the available options).

Then install the library.

1. Clone or download the source code.
2. Install:

    ```bash
    cctbx.python setup.py install
    ```

# Run

```bash
    conda activate cctbx
    restraintlib <printer_format> <input_file> <output_file>
```

# Test

Execute `pytest` in the main directory

# Generate functional restraints models.

The scripts are compatible with python 2.7, however the models are seriazized
(pickeld) so it might be compatible with python 2.7 or different `scikit-learn` version.
It is not the big problem, it is possible to recreate the modes from raw data
(`restraintlib/data/combined_results.csv`). Install dependencies and execute:

```bash
python retrain_restraintlib.py
```

or when installed manually

```bash
cctbx.python retrain_restraintlib.py
```



changelog.txt
# 2023.2.1

Override sigma for ShelxPrinter

# 2022.8.1

Update scikit-learn dependency version to 1.0.X

# 2022.6.1

Update values for P-O3', P-O5', P-O3'-C3', P-O5'-C5' to be based on https://doi.org/10.1093/nar/gkw717, Supplementary materials, Table S2

# 2022.5.2

Fix issues with missing residue ordering when res_id containing iCode (5hr7)

# 2022.5.1

Fix issues with missing atoms when checking torsion angle conditions (5CDB)

# 2021.10.1

Fix genration of restraints for non-standard nucleotides (like G49 in 427d)

# 2021.7.8

Fix for python2 compatibility

# 2021.7.7

Add retrain script
Support only for python >3.6
For python 2.7 one need to retrain the model manually

# 2021.07.6

Test new release

# 2021.07.5

Change atom selection order for Phenix

# 2021.07.4

Refactor code and add cdl_nucleotides function to launcher.py

# 2021.07.3

Change atom selection order for Phenix

# 2021.07.2

Override sigma values for Buster (1.5 for angles and 0.020 for bonds) and for Phenix (3.0 for angles, 0.020 for bonds)
Use "change" option for Phenix printer, with few exceptions

# 2021.07.1

Add Buster Printer.

# 2019.11.1

Make code python3 compatible.
Use pickle instead of joblib.
Add script to recalculate functional models.

# 2019.10.5

Fix minor bugs.


restraintlib/lib/PO4.py

PO4_PDB_CODES = ['A', 'C', 'G', 'T', 'U', 'DA', 'DC', 'DG', 'DT', 'DU', 'IC', 'IG']

PO4_ATOM_NAMES = {
    'P': 'P',
    'OP1': 'OP1',
    'O1P': 'OP1',
    'OP2': 'OP2',
    'O2P': 'OP2',
    "O5'": "O5'",
    "O5*": "O5'",
    "O3'": "O3'",
    "O3*": "O3'",
    "C5'": "C5'",
    "C5*": "C5'",
    "C3'": "C3'",
    "C3*": "C3'",
}

PO4_ATOM_RES = {
    'P': 0,
    'OP1': 0,
    'OP2': 0,
    "O5'": 0,
    "O3'": -1,
    "C5'": 0,
    "C3'": -1,
}

PO4_REQUIRED_CONDITION = {
    ("P", "O3'", 2.0, 0, -1),
    ("P", "O5'", 2.0, 0, 0),
    ("P", "OP1", 2.0, 0, 0),
    ("P", "OP2", 2.0, 0, 0),
    ("C3'", "O3'", 2.0, -1, -1),
    ("C5'", "O5'", 2.0, 0, 0),
}

PO4_DISTANCE_MEASURE = {
    'measure': 'euclidean_angles',
    'restraint_names': ['aO1O2', 'aO1O3', 'aO1O5', 'aO2O3', 'aO2O5', 'aO3O5']
}

PO4_CONDITION_DISTANCE_MEASURE = {
    'measure': 'euclidean_angles',
    'restraint_names': ['tC3O3P4O5', 'tC5O5P4O3']
}

PO4_RESTRAINTS = [
    {
        "conditions": [
            ["torsion", "tC3O3P4O5", ["O5'", "P", "O3'", "C3'"], -66.636, 7.779],
            ["torsion", "tC5O5P4O3", ["O3'", "P", "O5'", "C5'"], 171.37, 14.971]
        ],
        "name": "PO4==AA_0",
        "restraints": [
            ["angle", "aO1O2", ["OP1", "P", "OP2"], 117.6, 1.2],
            ["angle", "aO1O3", ["OP1", "P", "O3'"], 106.2, 1.1],
            ["angle", "aO1O5", ["OP1", "P", "O5'"], 110.2, 1.3],
            ["angle", "aO2O3", ["OP2", "P", "O3'"], 112.2, 1.0],
            ["angle", "aO2O5", ["OP2", "P", "O5'"], 109.3, 0.9],
            ["angle", "aO3O5", ["O3'", "P", "O5'"], 99.9, 0.7],
            ["angle", "aP4O3C3", ["P", "O3'", "C3'"], 120.8, 1.1],
            ["angle", "aP4O5C5", ["P", "O5'", "C5'"], 120.3, 1.9],
            ["dist", "dO1P4", ["OP1", "P"], 1.487, 0.01],
            ["dist", "dO2P4", ["OP2", "P"], 1.483, 0.01],
            ["dist", "dO3P4", ["O3'", "P"], 1.580, 0.010],
            ["dist", "dO5P4", ["O5'", "P"], 1.603, 0.011],
            ["dist", "dO3C3", ["O3'", "C3'"], 1.422, 0.010],
            ["dist", "dO5C5", ["O5'", "C5'"], 1.428, 0.013]
        ]
    },
    {
        "conditions": [
            ["torsion", "tC3O3P4O5", ["O5'", "P", "O3'", "C3'"], 171.37, 14.971],
            ["torsion", "tC5O5P4O3", ["O3'", "P", "O5'", "C5'"], -66.636, 7.779]
        ],
        "name": "PO4==AA_1",
        "restraints": [
            ["angle", "aO1O2", ["OP1", "P", "OP2"], 117.6, 1.2],
            ["angle", "aO1O3", ["OP1", "P", "O3'"], 109.3, 0.9],
            ["angle", "aO1O5", ["OP1", "P", "O5'"], 112.2, 1.0],
            ["angle", "aO2O3", ["OP2", "P", "O3'"], 110.2, 1.3],
            ["angle", "aO2O5", ["OP2", "P", "O5'"], 106.2, 1.1],
            ["angle", "aO3O5", ["O3'", "P", "O5'"], 99.9, 0.7],
            ["angle", "aP4O3C3", ["P", "O3'", "C3'"], 120.3, 1.9],
            ["angle", "aP4O5C5", ["P", "O5'", "C5'"], 120.8, 1.1],
            ["dist", "dO1P4", ["OP1", "P"], 1.483, 0.01],
            ["dist", "dO2P4", ["OP2", "P"], 1.487, 0.01],
            ["dist", "dO3P4", ["O3'", "P"], 1.603, 0.011],
            ["dist", "dO5P4", ["O5'", "P"], 1.580, 0.010],
            ["dist", "dO3C3", ["O3'", "C3'"], 1.422, 0.010],
            ["dist", "dO5C5", ["O5'", "C5'"], 1.428, 0.013]
        ]
    },
    {
        "conditions": [
            ["torsion", "tC3O3P4O5", ["O5'", "P", "O3'", "C3'"], -171.37, 14.971],
            ["torsion", "tC5O5P4O3", ["O3'", "P", "O5'", "C5'"], 66.636, 7.779]
        ],
        "name": "PO4==AA_2",
        "restraints": [
            ["angle", "aO1O2", ["OP1", "P", "OP2"], 117.6, 1.2],
            ["angle", "aO1O3", ["OP1", "P", "O3'"], 110.2, 1.3],
            ["angle", "aO1O5", ["OP1", "P", "O5'"], 106.2, 1.1],
            ["angle", "aO2O3", ["OP2", "P", "O3'"], 109.3, 0.9],
            ["angle", "aO2O5", ["OP2", "P", "O5'"], 112.2, 1.0],
            ["angle", "aO3O5", ["O3'", "P", "O5'"], 99.9, 0.7],
            ["angle", "aP4O3C3", ["P", "O3'", "C3'"], 120.3, 1.9],
            ["angle", "aP4O5C5", ["P", "O5'", "C5'"], 120.8, 1.1],
            ["dist", "dO1P4", ["OP1", "P"], 1.487, 0.01],
            ["dist", "dO2P4", ["OP2", "P"], 1.483, 0.01],
            ["dist", "dO3P4", ["O3'", "P"], 1.603, 0.011],
            ["dist", "dO5P4", ["O5'", "P"], 1.580, 0.010],
            ["dist", "dO3C3", ["O3'", "C3'"], 1.422, 0.010],
            ["dist", "dO5C5", ["O5'", "C5'"], 1.428, 0.013]
        ]
    },
    {
        "conditions": [
            ["torsion", "tC3O3P4O5", ["O5'", "P", "O3'", "C3'"], 66.636, 7.779],
            ["torsion", "tC5O5P4O3", ["O3'", "P", "O5'", "C5'"], -171.37, 14.971]
        ],
        "name": "PO4==AA_3",
        "restraints": [
            ["angle", "aO1O2", ["OP1", "P", "OP2"], 117.6, 1.2],
            ["angle", "aO1O3", ["OP1", "P", "O3'"], 112.2, 1.0],
            ["angle", "aO1O5", ["OP1", "P", "O5'"], 109.3, 0.9],
            ["angle", "aO2O3", ["OP2", "P", "O3'"], 106.2, 1.1],
            ["angle", "aO2O5", ["OP2", "P", "O5'"], 110.2, 1.3],
            ["angle", "aO3O5", ["O3'", "P", "O5'"], 99.9, 0.7],
            ["angle", "aP4O3C3", ["P", "O3'", "C3'"], 120.8, 1.1],
            ["angle", "aP4O5C5", ["P", "O5'", "C5'"], 120.3, 1.9],
            ["dist", "dO1P4", ["OP1", "P"], 1.483, 0.01],
            ["dist", "dO2P4", ["OP2", "P"], 1.487, 0.01],
            ["dist", "dO3P4", ["O3'", "P"], 1.580, 0.010],
            ["dist", "dO5P4", ["O5'", "P"], 1.603, 0.011],
            ["dist", "dO3C3", ["O3'", "C3'"], 1.422, 0.010],
            ["dist", "dO5C5", ["O5'", "C5'"], 1.428, 0.013]
        ]
    },
    {
        "conditions": [
            ["torsion", "tC3O3P4O5", ["O5'", "P", "O3'", "C3'"], -69.896, 9.625],
            ["torsion", "tC5O5P4O3", ["O3'", "P", "O5'", "C5'"], -68.72, 8.686]
        ],
        "name": "PO4==AS_0",
        "restraints": [
            ["angle", "aO1O2", ["OP1", "P", "OP2"], 119.9, 1.6],
            ["angle", "aO1O3", ["OP1", "P", "O3'"], 104.5, 0.9],
            ["angle", "aO1O5", ["OP1", "P", "O5'"], 110.3, 0.8],
            ["angle", "aO2O3", ["OP2", "P", "O3'"], 111.5, 1.1],
            ["angle", "aO2O5", ["OP2", "P", "O5'"], 105.2, 0.8],
            ["angle", "aO3O5", ["O3'", "P", "O5'"], 104.2, 1.5],
            ["angle", "aP4O3C3", ["P", "O3'", "C3'"], 121.5, 3.0],
            ["angle", "aP4O5C5", ["P", "O5'", "C5'"], 121.6, 2.8],
            ["dist", "dO1P4", ["OP1", "P"], 1.484, 0.012],
            ["dist", "dO2P4", ["OP2", "P"], 1.478, 0.01],
            ["dist", "dO3P4", ["O3'", "P"], 1.599, 0.016],
            ["dist", "dO5P4", ["O5'", "P"], 1.601, 0.016],
            ["dist", "dO3C3", ["O3'", "C3'"], 1.438, 0.007],
            ["dist", "dO5C5", ["O5'", "C5'"], 1.437, 0.017]
        ]
    },
    {
        "conditions": [
            ["torsion", "tC3O3P4O5", ["O5'", "P", "O3'", "C3'"], -68.72, 8.686],
            ["torsion", "tC5O5P4O3", ["O3'", "P", "O5'", "C5'"], -69.896, 9.625]
        ],
        "name": "PO4==AS_1",
        "restraints": [
            ["angle", "aO1O2", ["OP1", "P", "OP2"], 119.9, 1.6],
            ["angle", "aO1O3", ["OP1", "P", "O3'"], 105.2, 0.8],
            ["angle", "aO1O5", ["OP1", "P", "O5'"], 111.5, 1.1],
            ["angle", "aO2O3", ["OP2", "P", "O3'"], 110.3, 0.8],
            ["angle", "aO2O5", ["OP2", "P", "O5'"], 104.5, 0.9],
            ["angle", "aO3O5", ["O3'", "P", "O5'"], 104.2, 1.5],
            ["angle", "aP4O3C3", ["P", "O3'", "C3'"], 121.6, 2.8],
            ["angle", "aP4O5C5", ["P", "O5'", "C5'"], 121.5, 3.0],
            ["dist", "dO1P4", ["OP1", "P"], 1.478, 0.01],
            ["dist", "dO2P4", ["OP2", "P"], 1.484, 0.012],
            ["dist", "dO3P4", ["O3'", "P"], 1.601, 0.016],
            ["dist", "dO5P4", ["O5'", "P"], 1.599, 0.016],
            ["dist", "dO3C3", ["O3'", "C3'"], 1.438, 0.007],
            ["dist", "dO5C5", ["O5'", "C5'"], 1.437, 0.017]
        ]
    },
    {
        "conditions": [
            ["torsion", "tC3O3P4O5", ["O5'", "P", "O3'", "C3'"], 68.72, 8.686],
            ["torsion", "tC5O5P4O3", ["O3'", "P", "O5'", "C5'"], 69.896, 9.625]
        ],
        "name": "PO4==AS_2",
        "restraints": [
            ["angle", "aO1O2", ["OP1", "P", "OP2"], 119.9, 1.6],
            ["angle", "aO1O3", ["OP1", "P", "O3'"], 110.3, 0.8],
            ["angle", "aO1O5", ["OP1", "P", "O5'"], 104.5, 0.9],
            ["angle", "aO2O3", ["OP2", "P", "O3'"], 105.2, 0.8],
            ["angle", "aO2O5", ["OP2", "P", "O5'"], 111.5, 1.1],
            ["angle", "aO3O5", ["O3'", "P", "O5'"], 104.2, 1.5],
            ["angle", "aP4O3C3", ["P", "O3'", "C3'"], 121.6, 2.8],
            ["angle", "aP4O5C5", ["P", "O5'", "C5'"], 121.5, 3.0],
            ["dist", "dO1P4", ["OP1", "P"], 1.484, 0.012],
            ["dist", "dO2P4", ["OP2", "P"], 1.478, 0.01],
            ["dist", "dO3P4", ["O3'", "P"], 1.601, 0.016],
            ["dist", "dO5P4", ["O5'", "P"], 1.599, 0.016],
            ["dist", "dO3C3", ["O3'", "C3'"], 1.438, 0.007],
            ["dist", "dO5C5", ["O5'", "C5'"], 1.437, 0.017]
        ]
    },
    {
        "conditions": [
            ["torsion", "tC3O3P4O5", ["O5'", "P", "O3'", "C3'"], 69.896, 9.625],
            ["torsion", "tC5O5P4O3", ["O3'", "P", "O5'", "C5'"], 68.72, 8.686]
        ],
        "name": "PO4==AS_3",
        "restraints": [
            ["angle", "aO1O2", ["OP1", "P", "OP2"], 119.9, 1.6],
            ["angle", "aO1O3", ["OP1", "P", "O3'"], 111.5, 1.1],
            ["angle", "aO1O5", ["OP1", "P", "O5'"], 105.2, 0.8],
            ["angle", "aO2O3", ["OP2", "P", "O3'"], 104.5, 0.9],
            ["angle", "aO2O5", ["OP2", "P", "O5'"], 110.3, 0.8],
            ["angle", "aO3O5", ["O3'", "P", "O5'"], 104.2, 1.5],
            ["angle", "aP4O3C3", ["P", "O3'", "C3'"], 121.5, 3.0],
            ["angle", "aP4O5C5", ["P", "O5'", "C5'"], 121.6, 2.8],
            ["dist", "dO1P4", ["OP1", "P"], 1.478, 0.01],
            ["dist", "dO2P4", ["OP2", "P"], 1.484, 0.012],
            ["dist", "dO3P4", ["O3'", "P"], 1.599, 0.016],
            ["dist", "dO5P4", ["O5'", "P"], 1.601, 0.016],
            ["dist", "dO3C3", ["O3'", "C3'"], 1.438, 0.007],
            ["dist", "dO5C5", ["O5'", "C5'"], 1.437, 0.017]
        ]
    }
]
